Allow to get string references to strings inside torch::List

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39763 Allow to get string references to strings inside torch::List**

This is an ask from fluent. For performance reasons, they need a way to get read access to the std::string inside of a torch::List<std::string> without having to copy that string.

Instead of special casing std::string, we decided to give access to the underlying value. The API now looks like:

```cpp
torch::List<std::string> list = ...;
const std::string& str = list[2].toIValueRef().toStringRef();
```

Differential Revision: [D21966183](https://our.internmc.facebook.com/intern/diff/D21966183/)

smessmer committed:
Allow to get string references to strings inside torch::List

This is an ask from fluent. For performance reasons, they need a way to get read access to the std::string inside of a torch::List<std::string> without having to copy that string.

Instead of special casing std::string, we decided to give access to the underlying value. The API now looks like:

```cpp
torch::List<std::string> list = ...;
const std::string& str = list[2].toIValueRef().toStringRef();
```

Differential Revision: [D21966183](https://our.internmc.facebook.com/intern/diff/D21966183/)

[ghstack-poisoned]

ezyang reviewed (approved):
Uhh... sure, I guess? Seems a bit dodgy....

bhosmer commented:
IIRC on discussion @smessmer planned changes to this API, avoiding the indirection through IValue. Sebastian am I remembering where we wound up accurately?

smessmer commented:
yes, I'm adding a `.toStringRef()` directly to `ListElementReference`.

smessmer committed:
Update on "Allow to get string references to strings inside torch::List"


This is an ask from fluent. For performance reasons, they need a way to get read access to the std::string inside of a torch::List<std::string> without having to copy that string.

Instead of special casing std::string, we decided to give access to the underlying value. The API now looks like:

```cpp
torch::List<std::string> list = ...;
const std::string& str = list[2].toIValueRef().toStringRef();
```

Differential Revision: [D21966183](https://our.internmc.facebook.com/intern/diff/D21966183/)

[ghstack-poisoned]

smessmer committed:
Update on "Allow to get string references to strings inside torch::List"


This is an ask from fluent. For performance reasons, they need a way to get read access to the std::string inside of a torch::List<std::string> without having to copy that string.

Instead of special casing std::string, we decided to give access to the underlying value. The API now looks like:

```cpp
torch::List<std::string> list = ...;
const std::string& str = list[2].toIValueRef().toStringRef();
```

Differential Revision: [D21966183](https://our.internmc.facebook.com/intern/diff/D21966183/)

[ghstack-poisoned]
